notebooks/features/simple_deep_learning/DeepLearning - Deep Vision Classifier.ipynb

```diff
@@ -0,0 +1,182 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pyspark.sql.functions import udf, col\n",
+    "from pyspark.sql.types import IntegerType\n",
+    "\n",
+    "from pyspark.sql.functions import udf\n",
+    "\n",
+    "from pyspark.ml.evaluation import MulticlassClassificationEvaluator"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Read Dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def assign_label(path):\n",
+    "    num = int(path.split(\"/\")[-1].split(\".\")[0].split(\"_\")[1])\n",
+    "    return num // 81\n",
+    "\n",
+    "\n",
+    "assign_label_udf = udf(assign_label, IntegerType())"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# These files are already uploaded for build test machine\n",
+    "train_df = (\n",
+    "    spark.read.format(\"binaryFile\")\n",
+    "    .option(\"pathGlobFilter\", \"*.jpg\")\n",
+    "    .load(\"/tmp/17flowers/train\")\n",
+    "    .withColumn(\"features\", col(\"path\"))\n",
+    "    .withColumn(\"label\", assign_label_udf(col(\"path\")))\n",
+    "    .select(\"features\", \"label\")\n",
+    ")\n",
+    "\n",
+    "display(train_df.limit(100))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "test_df = (\n",
+    "    spark.read.format(\"binaryFile\")\n",
+    "    .option(\"pathGlobFilter\", \"*.jpg\")\n",
+    "    .load(\"/tmp/17flowers/test\")\n",
+    "    .withColumn(\"features\", col(\"path\"))\n",
+    "    .withColumn(\"label\", assign_label_udf(col(\"path\")))\n",
+    "    .select(\"features\", \"label\")\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from horovod.spark.common.store import DBFSLocalStore\n",
+    "from pytorch_lightning.callbacks import ModelCheckpoint\n",
+    "from synapse.ml.dl import *\n",
+    "\n",
+    "run_output_dir = \"/dbfs/FileStore/test/resnet50\"\n",
+    "store = DBFSLocalStore(run_output_dir)\n",
+    "epochs = 10\n",
+    "callbacks = [ModelCheckpoint(filename=\"{epoch}-{train_loss:.2f}\")]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "deep_vision_classifier = DeepVisionClassifier(\n",
+    "    backbone=\"resnet50\",\n",
+    "    store=store,\n",
+    "    callbacks=callbacks,\n",
+    "    num_classes=17,\n",
+    "    batch_size=16,\n",
+    "    epochs=epochs,\n",
+    "    validation=0.1,\n",
+    ")\n",
+    "\n",
+    "deep_vision_model = deep_vision_classifier.fit(train_df).setOutputCols([\"label_prob\"])"
```

Comment:
is setOutputCols the right api here? I think for a predictor there are explicit cols for probabilities, rawProbabilities, and predictions

Reply:
Yes, TorchModel extends HasOutputCol, but we can actually delete this setOutputCols as it internally gives it default values, and I also added predictionCol for our model. We couldn't follow the exact same behavior as predictor because the definition and rawProbabilities don't really apply here, but I changed the _transform function to add predictionCol so the output dataframe is more clear.